rpc/websocket.go

```diff
@@ -120,6 +120,56 @@ func (e wsHandshakeError) Error() string {
 	return s
 }
 
+func originIsAllowed(allowedOrigins mapset.Set, browserOrigin string) bool {
+	it := allowedOrigins.Iterator()
+	for origin := range it.C {
+		if ruleAllowsOrigin(origin.(string), browserOrigin) {
+			return true
+		}
+	}
+	return false
+}
+
+func ruleAllowsOrigin(allowedOrigin string, browserOrigin string) bool {
+	allowedScheme, allowedHostname, allowedPort, parsedAllowedErr := parseOriginURL(allowedOrigin)
+	browserScheme, browserHostname, browserPort, browserOriginErr := parseOriginURL(browserOrigin)
+	if parsedAllowedErr == nil && browserOriginErr == nil {
+		if allowedScheme != "" && allowedScheme != browserScheme {
+			return false
+		}
+		if allowedHostname != "" && allowedHostname != browserHostname {
+			return false
+		}
+		if allowedPort != "" && allowedPort != browserPort {
+			return false
+		}
+		return true
+	} else { // parse error
+		log.Warn("Error parsing Origin URL in comparison for match.", "allowedOrigin", allowedOrigin, "browserOrigin", browserOrigin)
+		return false
+	}
+}
+
+func parseOriginURL(origin string) (string, string, string, error) {
+	parsedURL, parseError := url.Parse(strings.ToLower(origin))
+	var scheme, hostname, port string
+	if parseError == nil {
```

Comment:
I'm not sure about using url.Parse here. The parser behaves differently depending on the four cases we want for the allow pattern (bare hostname, hostname+port, scheme+hostname+port, scheme+hostname). I made [this playground example](https://play.golang.org/p/VnQdfY2M6_t) to check where it puts things. Might be better to just use a regular expression here.

Reply:
It consolidates down to two cases based on whether or not the scheme is present.
If the scheme is present, then scheme/hostname/port are where you would expect them after url.Parse.
If not, then scheme is in the hostname slot and the port is in the hostname slot.  
The code in this PR handles those two like that.

Reply:
Clarification: In the case where it's just a bare hostname, that's in the "path" slot; that is a third case but still handled by the existing code in the PR (see lines just below this).